DOCS-6735 ACM: Dec 7 release (#122)
mulesoft/docs-api-community-manager: modules/ROOT/pages/data-bridge.adoc

= Set Up the Data Bridge
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images/

The data bridge is the secure connection from Anypoint API Community Manager to your Anypoint Platform organization.

API Community Manager components and data objects use Anypoint Platform as their external data source.

> 

Reviewer: So...now that your audience knows what this is, can you provide when/why they should set up the bridge and what the subtasks covered within this topic set are?

== Create a Connected App

> 

Reviewer: Is this the first step to setting up the bridge? Provide context regarding the containing task.

You can perform this step by using either the user interface or an API call.

> Your account must have the Org Admin permission.

Reviewer: Suggest providing a context for this task and then combining these separate lines into a following para:

"To create a connected app, you must have the Org Admin permission. You can create the app by using either the user interface or an API call."

=== Create a Connected App by Using the User Interface [[create-a-connected-app-with-the-user-interface]]

> To create a connected app by using the user interface, in Anypoint Platform, navigate to *Access Management* > *Connected Apps* > *Create app*, and provide this information:

Reviewer: ```suggestion
To create a connected app through the user interface, in Anypoint Platform, navigate to *Access Management* > *Connected Apps* > *Create app*, and take these actions:
```

> . Enter a value for the app *Name*.

Reviewer: ```suggestion
. Enter a value for *Name*.
```

. Set *Type* to *App acts on behalf of a user*.
. Set *Grant types* to *Authorization Code* and *Refresh Token*.

> . Set the mandatory *Website URL* to a URL of your choice.

Reviewer: "mandatory"? are the other settings not mandatory?

> . Set *Redirect URIs* to `+https://<domain>.my.salesforce.com/services/authcallback/Anypoint+` and replace `<domain>` with your domain.

Reviewer: ```suggestion
. Set *Redirect URIs* to `+https://<domain>.my.salesforce.com/services/authcallback/Anypoint+` and replace `<domain>` with your domain name.
```

. Set *Who can use this application?* to *Members of this organization only*.
. Click *Add Scopes*.
+
Select *Full Access* and *Background Access*.

> +

Reviewer: . Click *Add Scopes* and select *Full Access* and *Background Access*.

. Click *Create Application*.

> . Note the client ID and client secret.

Reviewer: ```suggestion
. Take note of the client ID and client secret for future use.
```

=== Create a Connected App by Using an API Call [[create-a-connected-app-with-an-api-call]]

> 

Reviewer: Are the examples that follow separate methods of doing the same thing? 
If so, let the audience know that there are different examples/subtasks here and not just one?

> To create a connected app by using an API call, follow this example using the `cURL` command.

Reviewer: ```suggestion
To create a connected app by using an API call, follow this example using the `cURL` command:
```

. Get an authorization token from Anypoint Platform:
+
----
    curl --silent https://anypoint.mulesoft.com/accounts/login -XPOST \
      -d 'username=<anypoint username>&password=<anypoint password>'
----
+

> . The command returns a response like this:

Reviewer: This is not a step.

+
----
    {
      "access_token": "5242XXXX-XXXX-XXXX-XXXX-XXXXXXXXfec7",
      "token_type": "bearer",
      "redirectUrl": "/home/"
    }
----
+

> . Note the `access_token` value.

Reviewer: GLOBALLY tweak these "Note <this>" texts to make clear that the user must write down or otherwise remember the information that is currently displayed because they will need it later.

+

> . Use the next `cURL` example to register an external client in Anypoint Platform for OpenID authentication.

Reviewer: This is formatted as a step but looks like a separate example?
Also, we should try to put the sentence that introduces a new element (list, graphic, table, syntax example...) directly before that element, and move any other info to before the intro sentence or after the new element.

+
Replace the access token in the authorization header with the `access_token` value.
+

> Replace `[YourOrgDomain]` with your Salesforce organization domain name. Find the domain name by logging in to your Salesforce organization, clicking *Setup*, and clicking the section *My Domain*.

Reviewer: These read like steps but are formatted as plain text, and they come between the text that introduces the syntax example.

      https://anypoint.mulesoft.com/accounts/api/connectedApplications \
      -H 'Content-Type: application/json' \
      -H 'Accept: */*' \
      -H 'Accept-Encoding: gzip, deflate' \
      -H 'Authorization: Bearer 5242XXXX-XXXX-XXXX-XXXX-XXXXXXXXfec7' \
      -H 'Host:anypoint.mulesoft.com' \
      -d '{"client_name":"ACM Integration App","grant_types":["authorization_code","refresh_token"],"redirect_uris":["https://[YourOrgDomain].my.salesforce.com/services/authcallback/Anypoint"],"scopes":["full","offline_access"],"public_keys":[]}'
----
+
. The `cURL` command returns a response like this:
+
----
    {
      "client_id": "5fafXXXXXXXXXXXXXXXXXXXXXXXX29c9",
      "client_secret": "9509XXXXXXXXXXXXXXXXXXXXXXXXC10E",
      "client_name": "ACM Integration App",
      "redirect_uris": [
        "https://[YourOrgDomain].my.salesforce.com/services/authcallback/Anypoint"
      ],
      "grant_types": [
        "authorization_code",
        "refresh_token"
      ],
      "public_keys": [],
      "scopes": [
        "offline_access",
        "full"
      ],
      "enabled": true,
      "owner_org_id": "f377XXXX-XXXX-XXXX-XXXX-XXXXXXXX9d08",
      "as_id": "anypoint"
    }
----
+

> . Note the `client_id` and `client_secret` values.

Reviewer: ditto re. "Note...."

== Create an Authentication Provider in Your Salesforce Organization [[create-auth-provider]]

> NOTE: If your Anypoint Platform organization is hosted in the EU, change `+https://anypoint.mulesoft.com+` to `+https://eu1.anypoint.mulesoft.com+`.

Reviewer: This looks like info specific to one or more steps below; should it appear with those steps?
Or maybe this should be plain text introducing the task and referring to the whole task:
```suggestion
In the following task steps, if your Anypoint Platform organization is hosted in the EU, replace all instances of `+https://anypoint.mulesoft.com+` with `+https://eu1.anypoint.mulesoft.com+`.
```

Author: Yes, I think the plain text introducing the task and referring to the whole task is the best option.

> . In *Setup*, search for *Auth. Providers* using the *Quick Find* box and click *Auth. Providers* -> *New*.

Reviewer: We've had a lot of multiaction steps so far, which is kind of a no-no, but this one bothers me more than most:
```suggestion
. In *Setup*, search for *Auth. Providers* using the *Quick Find* box.
. Click *Auth. Providers* -> *New*.
```

> . In *Provider type* select *Open ID Connect*.

Reviewer: ```suggestion
. In *Provider type*, select *Open ID Connect*.
```

. Set *Name* to `Anypoint`.

> . Set *Consumer Key* to the application ID of the connected app created in Anypoint Platform.

Reviewer: ```suggestion
. Set *Consumer Key* to the application ID of the connected app (which you took note of when you created the app in Anypoint Platform).
```

+

> This ID was returned by the Anypoint UI at the end of the task <<create-a-connected-app-with-the-user-interface>> or returned by the Anypoint API call at the end of the task <<create-a-connected-app-with-an-api-call>>.

Reviewer: More complicated than suggestion, I think....
```suggestion
```

+
. Set *Consumer Secret* to the application secret of the connected app created in Anypoint Platform.
+
This secret was returned by the Anypoint UI at the end of the task <<create-a-connected-app-with-the-user-interface>> or returned by the Anypoint API call at the end of the task <<create-a-connected-app-with-an-api-call>>.

> +

Reviewer: ditto previous
If we make any refs. at all, I suggest we ref. FORWARD, at the point that the ID and secret are first received, saying that the user needs them for <this task>, instead of referring BACKWARD to that time here.

. Set *Authorize Endpoint URL* to `+https://anypoint.mulesoft.com/accounts/api/v2/oauth2/authorize+` .
. Set *Token Endpoint URL* to `+https://anypoint.mulesoft.com/accounts/api/v2/oauth2/token+` .
. Set *Default Scopes* to `full offline_access`.
. Click *Save*.

> NOTE: Remember to associate the external data source with this authentication provider.

Reviewer: WHEN does the user do this: now? Should that be the final step, then?

== Authenticate the Data Bridge [[authenticate-data-bridge]]

> 

Reviewer: Provide introduction, with context

. Search for *External Data Source* in the *Quick Find* box and then click *External Data Sources*.

> . Locate the *Exchange* item and then click *Edit*.

Reviewer: ```suggestion
. Locate *Exchange*  and click *Edit*.
```

. Update *Identity Type* to *Named Principal*.
. Update *Authentication Protocol* to *OAuth 2.0*.
. Update *Authentication Provider* to *Anypoint*.
. Click *Save*.

> . If you don't have an active session in your Anypoint Platform organization, the *Anypoint Login* page is shown. Log in to Anypoint Platform to complete the authentication. If you already have an active session, authentication is performed without requiring a login.

Reviewer: ```suggestion
. If you don't have an active session in your Anypoint Platform organization, log in.
```

> . Verify the Exchange *External Data Sources* *Authentication Status* is *Authenticated*.

Reviewer: ```suggestion
. Verify that the authentication status of the Exchange *External Data Sources* is *Authenticated*.
```
"the Exchange *External Data Sources* what: section? pane? field? What kind of control does this label refer to?

> After authenticating the External Data Source, provide the credentials of an Anypoint Platform system user with read-only access, for Lightning components like the API Console which retrieve data directly from the Anypoint Platform APIs.

Reviewer: ```suggestion
After you authenticate the external data source, you next provide the credentials of an Anypoint Platform system user with read-only access for Lightning components, like the API Console, which retrieve data directly from the Anypoint Platform APIs:
```
This is an intro to the following steps, right?

> . Obtain the username and password for the system user with read-only privileges in your Anypoint Platform organization.

Reviewer: This sentence can be read in a couple of different ways.
Is the step to obtain the username and password for the system user to subsequently use?
Or to obtain the username and password of that type of user?

> . Search for *Named Credentials* in the *Quick Find* box and then click *Named Credentials*.

Reviewer: ```suggestion
. In the *Quick Find* box, click *Named Credentials*.
```

> . Locate the *anypoint.mulesoft* item and then click *Edit*.

Reviewer: ```suggestion
. Locate *anypoint.mulesoft* and click *Edit*.
```

> . In the *Authentication* section, enter the Anypoint Platform username and password for the read-only system user in the *Username* and

Reviewer: ```suggestion
. In the *Authentication* section, enter the Anypoint Platform username and password for the read-only system user in the *Username* and *Password* fields.
```

> *Password* fields.

Reviewer: delete
```suggestion
```
